src/adapter/src/coord/statement_logging.rs

```diff
@@ -43,13 +46,116 @@ pub enum PreparedStatementLoggingInfo {
+    /// prepared statement. Possibly mutates the `PreparedStatementLoggingInfo` metadata.
+    ///
+    /// This function does not do a sampling check, and assumes we did so in a higher layer.
+    pub(crate) fn log_prepared_statement(
         &mut self,
         session: &mut Session,
         logging: &Arc<QCell<PreparedStatementLoggingInfo>>,
+    ) -> (Option<StatementPreparedRecord>, Uuid) {
+        let logging = session.qcell_rw(&*logging);
+        let mut out = None;
+
+        let uuid = match logging {
+            PreparedStatementLoggingInfo::AlreadyLogged { uuid } => *uuid,
+            PreparedStatementLoggingInfo::StillToLog {
+                sql,
+                prepared_at,
+                name,
+                session_id,
+                accounted,
+            } => {
+                assert!(
+                    *accounted,
+                    "accounting for logging should be done in `begin_statement_execution`"
+                );
+                let uuid = Uuid::new_v4();
+                out = Some(StatementPreparedRecord {
+                    id: uuid,
+                    sql: std::mem::take(sql),
+                    name: std::mem::take(name),
+                    session_id: *session_id,
+                    prepared_at: *prepared_at,
+                });
+
+                *logging = PreparedStatementLoggingInfo::AlreadyLogged { uuid };
+                uuid
+            }
+        };
+        (out, uuid)
+    }
+    /// The rate at which statement execution should be sampled.
+    /// This is the value of the session var `statement_logging_sample_rate`,
+    /// constrained by the system var `statement_logging_max_sample_rate`.
+    pub fn statement_execution_sample_rate(&self, session: &Session) -> f64 {
+        let system: f64 = self
+            .catalog()
+            .system_config()
+            .statement_logging_max_sample_rate()
+            .try_into()
+            .expect("value constrained to be convertible to f64");
+        let user: f64 = session
+            .vars()
+            .get_statement_logging_sample_rate()
+            .try_into()
+            .expect("value constrained to be convertible to f64");
+        f64::min(system, user)
+    }
+
+    /// Record the end of statement execution for a statement whose beginning was logged.
+    /// It is an error to call this function for a statement whose beginning was not logged
+    /// (because it was not sampled). Requiring the opaque `StatementLoggingId` type,
+    /// which is only instantiated by `begin_statement_execution` if the statement is actually logged,
+    /// should prevent this.
+    pub fn end_statement_execution(
+        &mut self,
+        StatementLoggingId(id): StatementLoggingId,
+        reason: StatementEndedExecutionReason,
     ) {
+        let now = self.now_datetime();
+        let now_millis = now.timestamp_millis().try_into().expect("sane system time");
```

Comment:
Instead of taking a timestamp here should we move this into `ExecuteContext::retire` and add it to `Message::RetireExecute`? Since `Message::RetireExecute` is put into a channel and processed sometime after the fact these timestamps might not be very accurate to the time at which retire was called

Reply:
That's a good point. I suppose it's ambiguous -- has a statement "finished" once control is returned to the user, or once the coordinator has cleaned up the state related to it? I'm leaning towards implementing your suggestion, though I don't think it's 100% clear.